src/persist/src/indexed/encoding.rs

```diff
@@ -151,7 +151,61 @@ pub struct BlobTraceBatchPart<T> {
     /// Index of this part in the list of parts that form the batch.
     pub index: u64,
     /// The updates themselves.
-    pub updates: Vec<ColumnarRecords>,
+    pub updates: BlobTraceUpdates,
+}
+
+/// The set of updates that are part of a [`BlobTraceBatchPart`].
+#[derive(Clone, Debug)]
+pub enum BlobTraceUpdates {
+    /// Legacy format. Keys and Values are encoded into bytes via [`Codec`], then stored in our own
+    /// columnar-esque struct.
+    /// 
+    /// [`Codec`]: mz_persist_types::Codec
+    Row(Vec<ColumnarRecords>),
+    /// Migration format. Keys and Values are encoded into bytes via [`Codec`] and structured into
+    /// an Apache Arrow columnar format.
+    /// 
+    /// [`Codec`]: mz_persist_types::Codec
+    Both((ColumnarRecords, ColumnarRecordsStructuredExt)),
+    // TODO(parkmycar): Write only columnar/Arrow data.
+}
+
+impl BlobTraceUpdates {
+    /// Return the [`ColumnarRecords`], if one exists for the provided `idx`.
+    pub fn get(&self, idx: usize) -> Option<&ColumnarRecords> {
+        match self {
+            BlobTraceUpdates::Row(updates) => updates.get(idx),
+            BlobTraceUpdates::Both((codec, _structured)) => match idx {
+                0 => Some(codec),
+                _ => None,
+            },
+        }
+    }
+
+    /// Returns the number of updates.
```

Comment:
is this doc correct? looks like it returns the number of `ColumnarRecords` chunks, which each could have any number of updates

edit: oh, I see. the name of this method is confusing IMO. `num_row_groups` or something?

Reply:
Updated the name to be `num_row_groups`